### PR TITLE
[RISCV] Add regalloc hints for Zcb instructions.

### PR DESCRIPTION
**Pull request by topperc**
This hints the register allocator to use the same register for source and destination to enable more compression.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp [hunk 1]
```diff
@@ -789,9 +790,16 @@ bool RISCVRegisterInfo::getRegAllocationHints(
...
-    case RISCV::ANDI:
+    case RISCV::ANDI: {
       NeedGPRC = true;
-      return MI.getOperand(2).isImm() && isInt<6>(MI.getOperand(2).getImm());
+      if (!MI.getOperand(2).isImm())
+        return false;
+      int64_t Imm = MI.getOperand(2).getImm();
+      if (isInt<6>(Imm))
+        return true;
+      // c.zext.h
```
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp:800]:
```suggestion
      // c.zext.b
```

File: llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp [hunk 2]
```diff
@@ -802,6 +810,23 @@ bool RISCVRegisterInfo::getRegAllocationHints(
...
+    case RISCV::SEXT_H:
+    case RISCV::ZEXT_H_RV32:
+    case RISCV::ZEXT_H_RV64:
+      // c.mul, c.sext.b, c.sext.h, c.zext.h
+      NeedGPRC = true;
+      return Subtarget.hasStdExtZcb();
+    case RISCV::ADD_UW:
+      // c.zext.w
+      NeedGPRC = true;
+      return Subtarget.hasStdExtZcb() && MI.getOperand(2).getReg() == RISCV::X0;
```
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVRegisterInfo.cpp:824]:
```suggestion
      return Subtarget.hasStdExtZcb() && MI.getOperand(2).isReg() && MI.getOperand(2).getReg() == RISCV::X0;
```

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thanks!

**Review by asb — APPROVED** [review]
LGTM.

**Pull request merged**